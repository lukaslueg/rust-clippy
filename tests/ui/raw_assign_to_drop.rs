#![warn(clippy::raw_assign_to_drop)]

fn main() {
    unsafe fn foo(r: *mut String, i: *mut i32) {
        *r = "foo".to_owned();

        // no lint on {integer}
        *i = 47;

        (*r, *r) = ("foo".to_owned(), "bar".to_owned());

        (*r, *i) = ("foo".to_owned(), 47);

        let mut x: String = Default::default();
        *(&mut x as *mut _) = "Foo".to_owned();

        // no lint on `u8`
        *x.as_mut_ptr() = b'a';

        let mut v: Vec<String> = vec![];
        *v.as_mut_ptr() = Default::default();
    }
}
