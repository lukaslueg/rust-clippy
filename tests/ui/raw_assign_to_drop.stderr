error: assignment via raw pointer always executes destructor
  --> tests/ui/raw_assign_to_drop.rs:5:10
   |
LL |         *r = "foo".to_owned();
   |          ^ this place may be uninitialized, causing Undefined Behavior when the destructor executes
   |
   = note: the destructor defined by `std::string::String` is executed during assignment of the new value
   = help: use `std::ptr::write()` to overwrite a (possibly uninitialized) place
   = help: use `std::ptr::drop_in_place()` to drop the previous value if such value exists
   = note: `-D clippy::raw-assign-to-drop` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::raw_assign_to_drop)]`

error: assignment via raw pointer always executes destructor
  --> tests/ui/raw_assign_to_drop.rs:10:11
   |
LL |         (*r, *r) = ("foo".to_owned(), "bar".to_owned());
   |           ^ this place may be uninitialized, causing Undefined Behavior when the destructor executes
   |
   = note: the destructor defined by `std::string::String` is executed during assignment of the new value
   = help: use `std::ptr::write()` to overwrite a (possibly uninitialized) place
   = help: use `std::ptr::drop_in_place()` to drop the previous value if such value exists

error: assignment via raw pointer always executes destructor
  --> tests/ui/raw_assign_to_drop.rs:10:15
   |
LL |         (*r, *r) = ("foo".to_owned(), "bar".to_owned());
   |               ^ this place may be uninitialized, causing Undefined Behavior when the destructor executes
   |
   = note: the destructor defined by `std::string::String` is executed during assignment of the new value
   = help: use `std::ptr::write()` to overwrite a (possibly uninitialized) place
   = help: use `std::ptr::drop_in_place()` to drop the previous value if such value exists

error: assignment via raw pointer always executes destructor
  --> tests/ui/raw_assign_to_drop.rs:12:11
   |
LL |         (*r, *i) = ("foo".to_owned(), 47);
   |           ^ this place may be uninitialized, causing Undefined Behavior when the destructor executes
   |
   = note: the destructor defined by `std::string::String` is executed during assignment of the new value
   = help: use `std::ptr::write()` to overwrite a (possibly uninitialized) place
   = help: use `std::ptr::drop_in_place()` to drop the previous value if such value exists

error: assignment via raw pointer always executes destructor
  --> tests/ui/raw_assign_to_drop.rs:15:10
   |
LL |         *(&mut x as *mut _) = "Foo".to_owned();
   |          ^^^^^^^^^^^^^^^^^^ this place may be uninitialized, causing Undefined Behavior when the destructor executes
   |
   = note: the destructor defined by `std::string::String` is executed during assignment of the new value
   = help: use `std::ptr::write()` to overwrite a (possibly uninitialized) place
   = help: use `std::ptr::drop_in_place()` to drop the previous value if such value exists

error: assignment via raw pointer always executes destructor
  --> tests/ui/raw_assign_to_drop.rs:21:10
   |
LL |         *v.as_mut_ptr() = Default::default();
   |          ^^^^^^^^^^^^^^ this place may be uninitialized, causing Undefined Behavior when the destructor executes
   |
   = note: the destructor defined by `std::string::String` is executed during assignment of the new value
   = help: use `std::ptr::write()` to overwrite a (possibly uninitialized) place
   = help: use `std::ptr::drop_in_place()` to drop the previous value if such value exists

error: aborting due to 6 previous errors

